geps/gep-3155/index.md

```diff
@@ -0,0 +1,205 @@
+}
+```
+
+**2. Add a new `Client` field at the top level of BackendTLSPolicy**
+
+```go
+type BackendTLSPolicySpec {
+  // Client specifies overriden TLS settings for the given service.
+  //
+  // If Client is omitted, backend mTLS settings inherited from the given gateway are
+  // used.
+  Client TLSClientSettings `json:"client"`
+}
+
+type TLSClientSettings struct {
+  // CertificateRef is a reference to an object that contains a Client
+  // Certificate.
+  //
+  // References to a resource in different namespace are invalid UNLESS there
+  // is a ReferenceGrant in the target namespace that allows the certificate
+  // to be attached. If a ReferenceGrant does not allow this reference, the
+  // "ResolvedRefs" condition MUST be set to False for this listener with the
+  // "RefNotPermitted" reason.
+  //
+  // CertificateRef can reference to standard Kubernetes resources, i.e.
+  // Secret, or implementation-specific custom resources.
+  //
+  // If CertificateRef is not specified, no client certificate is presented when
+  // connecting to the service.
+  CertificateRef SecretObjectReference `json:"certificateRef,omitempty"`
+}
+```
+
+#### Limitations
+
+Configuring client certificate on the service level may result in multiple Gateways
+from different vendors sharing the same identity when connecting to a single service.
+
+The problem could be partially mitigated by allowing shared policies to selectively 
+target specific Gateway instances, GatewayClasses or gateway namespaces.
+
+### SANs on BackendTLSPolicy
+
+This change enables the certificate to have a different identity than the SNI
+(both are currently tied to the hostname field). This is particularly useful
+when using SPIFFE, which relies on URI Subject Names which are not valid SNIs 
+as per https://www.rfc-editor.org/rfc/rfc6066.html#section-3.
+
+In such case either connection properties or an arbitrary SNI, like cluster-local 
+service name could be used for certificate selection, while the identity validation
+will be done based on SubjectAltNames field.
+
+**1. Add a new `SubjectAltNames` field to `BackendTLSPolicyValidation`**
+
+```go
+type BackendTLSPolicyValidation struct {
+  // SubjectAltNames contains one or more Subject Alternative Names.
+  // When specified, at least one of certificate's Subject Alternate Names MUST 
+  // match at least one of the specified SubjectAltNames.
+  // +kubebuilder:validation:MaxItems=5
+  SubjectAltNames []SubjectAltName `json:"subjectAltNames,omitempty"`
+}
+
+type SubjectAltName struct {
+  // Hostname contains Subject Alternative Name specified in DNS name format. It is
+  // mutually exclusive with Hostname. At least one of the fields must be specified.
+  Hostname v1.PreciseHostname `json:"hostname"`
+
+  // URI contains Subject Alternative Name specified in URI format. It is mutually
+  // exclusive with Hostname. At least one of the fields must be specified.
+  URI string `json:"uri,omitempty"`
+}
```

Comment:
I think that this struct may be better handled as either a type/value struct, as follows:
```suggestion
type SubjectAltName struct {
  // Type chooses the type of SAN stored in this struct. Valid values are either `hostname`
  // or `uri`.
  Type string `json:"type"`
  
  // Value holds the value of the SAN for the type.
  Value string `json:"value"`
}
```

or as a type-selected union, like this:

```suggestion
type SubjectAltName struct {
  // Type chooses the type of SAN stored in this struct. Valid values are either `hostname`
  // or `uri`. The corresponding field must be filled out based on the type (the
  // Hostname or URI fields respectively). If a field not corresponding to the Type is
  // nonzero, it will be ignored.
  Type string `json:"type"`
  
  // Hostname contains Subject Alternative Name specified in DNS name format. It is
  // mutually exclusive with Hostname. At least one of the fields must be specified.
  Hostname v1.PreciseHostname `json:"hostname"`

  // URI contains Subject Alternative Name specified in URI format. It is mutually
  // exclusive with Hostname. At least one of the fields must be specified.
  URI string `json:"uri,omitempty"`
}
```

The first has the advantage of brevity, at the cost of some type-safety and the associated validation on ingestion - it would need more checking in-code instead. The second keeps the type-safety, and works the same as many other type-selected union fields in Kubernetes.

I probably slightly favor the second.

Reply:
Thanks, I will update to type-selected union.